File: 301_Remove_Invalid_Parentheses/301_1.java
class Solution {
    public List<String> removeInvalidParentheses(String s) {
        Set<String> level = new HashSet<>();
        level.add(s);
        while (true) {
            List<String> valid = new ArrayList<>();
            for (String str : level) {
                if (isValid(str)) {
                    valid.add(str);
                }
            }
            if (!valid.isEmpty()) return valid;
            Set<String> nextLevel = new HashSet<>();
            for (String item: level) {
                for (int i = 0; i < item.length(); i++) {
                    if (item.charAt(i) == '(' || item.charAt(i) == ')') {
                        nextLevel.add(item.substring(0, i) + item.substring(i + 1));
                    }
                }
            }
            level = nextLevel;
            if (level.isEmpty()) return new ArrayList<>();
        }
    }

    private boolean isValid(String str) {
        int count = 0;
        for (char c : str.toCharArray()) {
            if      (c == '(') count++;
            else if (c == ')') count--;
            if (count < 0) return false;
        }
        return count == 0;
    }
}

File: 301_Remove_Invalid_Parentheses/301_1.py
class Solution:
    def removeInvalidParentheses(self, s: str) -> List[str]:
        def isValid(string):
            count = 0
            for i in string:
                if i == '(':
                    count += 1
                elif i == ')':
                    count -= 1
                if count < 0:
                    return False
            return count == 0
        
        level = set([s])
        while True:
            valid = [i for i in level if isValid(i)]
            if valid:
                return valid
            next_level = set()
            for string in level:
                for i in range(len(string)):
                    if string[i] in "()":
                        next_level.add(string[:i] + string[i+1:])
            level = next_level
            if not level:
                return []
